### PR TITLE
Enhance shuffle-test of `testing.for_dtypes_combination`

### PR DESCRIPTION
**Pull request by asi1024**
Current shuffle-test sometimes causes stochastic test failure because the dtypes combination whose `result_type` is `bool` is rarely chosen. This PR prevents passing such tests.

### CODE REVIEW

**Comment by niboshi** [comment]
This PR in the current form introduces a possibility of duplicate combinations.
How about avoiding that?

File: cupy/testing/helper.py [hunk 1]
```diff
@@ -868,11 +874,12 @@ def for_dtypes_combination(types, names=('dtype',), full=None):
...
-            t = list(types)
-            random.shuffle(t)
-            ts.append(t)
+            shuffled_types = types[:]
+            random.shuffle(shuffled_types)
+            ts.append(types + shuffled_types)
 
-        combination = [dict(zip(names, typs)) for typs in zip(*ts)]
+        combination = [tuple(zip(names, typs)) for typs in zip(*ts)]
+        combination = [dict(assoc_list) for assoc_list in set(combination)]
```
Comment (niboshi) [inline cupy/testing/helper.py]: Let's add a comment.
```suggestion
        # Remove duplicate entries
        combination = [dict(assoc_list) for assoc_list in set(combination)]
```

**Comment by niboshi** [comment]
Please update the documentation as well.

**Comment by asi1024** [comment]
@niboshi Can I remove the documents about the details of the algorithm of this function? I don't want to guarantee how dtypes are chosen.

**Comment by niboshi** [comment]
@asi1024 OK, please do so.

File: cupy/testing/helper.py [hunk 2]
```diff
@@ -839,26 +839,18 @@ def for_dtypes_combination(types, names=('dtype',), full=None):
...
-    :math:`2^3` patterns. On the other hand, if ``full`` is ``False``,
-    shuffled lists are made for ``a_type`` and ``b_type``.
-    Suppose the lists are ``(16, 64, 32)`` for ``a_type`` and
-    ``(32, 64, 16)`` for ``b_type`` (prefixes are removed for short).
-    Then the combinations of ``(a_type, b_type)`` to be tested are
-    ``(16, 32)``, ``(64, 64)`` and ``(32, 16)``.
+    combinations is tested. If ``full`` is ``None``, the default
+    behavior is chosen. The default behavior is initially the same as
+    ``False``, but it is switched by setting the environment variable
+    ``CUPY_TEST_FULL_COMBINATION=1``.
```
Comment (niboshi) [inline cupy/testing/helper.py]: How about this wording?

"So, if ``full`` is ``False``, only a subset of possible combinations is randomly sampled. If ``full`` is ``None``, the behavior is determined by an environment variable ``CUPY_TEST_FULL_COMBINATION``. If the value is set to `'1'`, it behaves as if `full=True`, and otherwise `full=False`.
"

Comment (asi1024) [reply cupy/testing/helper.py]: Corrected the phrase, as you suggested. PTAL.

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
LGTM
